### PR TITLE
WIP: Scale ha combined

### PR DESCRIPTION
**Pull request by jayunit100**
Updated PR From #12830 .
Please close issue 12830 , I made a mistake when creating the PR.

### CODE REVIEW

**Comment by jayunit100** [comment]
@mikedanese @timothysc @rrati in here, we can continue.....
quick summary of where we were ... 
- Mike has suggested splitting into multiple PRs (2 or 3).  Im okay with that but slightly lean towards doing it as a monolith.
- Some json fields needed to be updated i think.
- Some cleanup needs to be done, also verflags is now unreachable.

**Comment by mikedanese** [comment]
All my comments apply from #12830. Carrying over:

> @jayunit100 I think we should divide this into multiple PRs that build on eachother. A logical division would be:
> 
> Lock/Lease API and resource in the experimental API (maybe include pkg/client and kubectl integration but that could also be separate)
> Refactor podmaster.go into a library that lives in pkg/lease that uses the Lock/Lease API rather than etcd to implement useful patterns.
> Retrofit controller-manager and scheduler to use pkg/lease
> ...
> This will make this chunk of work much easier to review and hopefully we can merge faster. WDYT?

**Comment by timothysc** [comment]
@mikedanese @jayunit100 - I'm ok with the breakup specified.

**Comment by jayunit100** [comment]
Ok. I'm testing failover now in multiple daemons will let you guys know how it looks in the morning

**Comment by bgrant0607** [comment]
Ref #1542

File: pkg/api/types.go [hunk 1]
```diff
@@ -2154,3 +2154,32 @@ type RangeAllocation struct {
 	// a single allocated address (the fifth bit on CIDR 10.0.0.0/8 is 10.0.0.4).
 	Data []byte `json:"data"`
 }
+
+// LockSpec is the specification of a resource lock.
+type LockSpec struct {
+	// Who currently has the lock.
+	HeldBy string `json:"heldby"`
+	// How long the lock is valid.
+	LeaseTime uint64 `json:"duration"`
```
Comment (wojtek-t) [inline pkg/api/types.go]: I think that for consistency with other objects, we should use "util.Time" isntead of uint64/string in all those places.

File: pkg/registry/lock/registry.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
+/*
```
Comment (wojtek-t) [inline pkg/registry/lock/registry.go]: Can you please be consistent with all the other resources:
- registry.go file contains "Registry" interface (Create/Update/Delete/List/Get/Watch methods)
- rest.go file contains "strategy" (if needed)
- etcd/etcd.go contains "REST" definition

I agree that it's very misleading (we are going to change it). But for consistency it's much better to have everything to be the same.

File: pkg/api/types.go (hunk 1 above)
Comment (jayunit100) [inline pkg/api/types.go]: Not a bad idea right now in lease.go I parse this value and its hacks... Also a comment on how and where ttl is implemented on renewal.

**Comment by mikedanese** [comment]
@jayunit100 or @rrati please open a PR that contains just the api addition as soon as you can. I want to minimize the number of review comments we lose when we switch PRs.

**Comment by bgrant0607** [comment]
cc @kubernetes/sig-scalability

**Comment by jayunit100** [comment]
Hi folks.  @timothysc and @rrati : Should I rebase these PRs accordingly?  I  can split branches up as needed by tomorrow (they are both out at mesoscon right now), so that we have three branches

- R: robs branch (locking).
  - L: branch which branches from his (lease.go) branching from R.
    - P: podmaster branch branching from L.
    - HA: ha branch branching from L (scheduler and controller manager mods).

Sounds ok? I can retain the commits as is in the process.

**Comment by mikedanese** [comment]
:+1: Looking forward to the Lease API PR.

**Comment by jayunit100** [comment]
Got ack from @timothysc whose at mesoscon he's ready to start separating branches. today
- I'll finish some final tests, then I'll restructure the branches as described above.  
- Then we will have to start the process of rebasing and so on.

**Comment by jayunit100** [comment]
MOVED AGAIN AND SPLIT INTO THE TWO CORE PRs.

#13037 (locking and leasing)
#13040 (scheduler, controller)

**Pull request closed without merging**